src/AppInstallerCLICore/PortableInstaller.cpp

```diff
@@ -0,0 +1,489 @@
+    {
+        if (scope == Manifest::ScopeEnum::Machine)
+        {
+            return Runtime::GetPathTo(Runtime::PathName::PortableLinksMachineLocation);
+        }
+        else
+        {
+            return Runtime::GetPathTo(Runtime::PathName::PortableLinksUserLocation);
+        }
+    }
+
+    std::filesystem::path GetPortableInstallRoot(Manifest::ScopeEnum scope, Utility::Architecture arch)
+    {
+        if (scope == Manifest::ScopeEnum::Machine)
+        {
+            if (arch == Utility::Architecture::X86)
+            {
+                return Runtime::GetPathTo(Runtime::PathName::PortablePackageMachineRootX86);
+            }
+            else
+            {
+                return Runtime::GetPathTo(Runtime::PathName::PortablePackageMachineRootX64);
+            }
+        }
+        else
+        {
+            return Runtime::GetPathTo(Runtime::PathName::PortablePackageUserRoot);
+        }
+    }
+
+    bool PortableInstaller::VerifyPortableFilesForUninstall()
+    {
+        const auto& indexPath = GetPortableIndexPath();
+        if (std::filesystem::exists(indexPath))
+        {
+            PortableIndex portableIndex = PortableIndex::Open(indexPath.u8string(), SQLiteStorageBase::OpenDisposition::ReadWrite);
+            std::vector<IPortableIndex::PortableFile> portableFiles = portableIndex.GetAllPortableFiles();
+            for (const auto& file : portableFiles)
+            {
+                IPortableIndex::PortableFileType fileType = file.FileType;
+                const auto& filePath = file.GetFilePath();
+
+                if (fileType == IPortableIndex::PortableFileType::File)
+                {
+                    if (std::filesystem::exists(filePath) && !SHA256::AreEqual(SHA256::ComputeHashFromFile(filePath), SHA256::ConvertToBytes(file.SHA256)))
+                    {
+                        return false;
+                    }
+                }
+                else if (fileType == IPortableIndex::PortableFileType::Symlink)
+                {
+                    if (Filesystem::SymlinkExists(filePath) && !Filesystem::VerifySymlink(filePath, file.SymlinkTarget))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+        else
+        {
+            if (std::filesystem::exists(PortableTargetFullPath) && !SHA256::AreEqual(SHA256::ComputeHashFromFile(PortableTargetFullPath), SHA256::ConvertToBytes(SHA256)))
+            {
+                return false;
+            }
+            else if (Filesystem::SymlinkExists(PortableSymlinkFullPath) && !Filesystem::VerifySymlink(PortableSymlinkFullPath, PortableTargetFullPath))
+            {
+                return false;
+            }
+            else
+            {
+                return true;
+            }
+        }
+    }
+
+    HRESULT PortableInstaller::InstallSingle(const std::filesystem::path& installerPath)

```

Comment:
I don't think splitting single/multiple is necessarily a good idea from a maintainability standpoint.  It isn't really necessary either given what I'm going to suggest in this rather long comment.

Any action (install/upgrade/uninstall) taken on the portable file(s) can be described by 3 states: { Desired, Expected, Actual }
- Desired contains the incoming file(s)
- Expected contains the description of the file(s) that we maintain (registry or index)
- Actual is the state of the filesystem

Every filesystem entry that we want can be described in the desired state, and each one has a related state in the expected and actual sets (even if it isn't in the set in memory).  This includes the root install directory for the package, which should not be treated as a special case.  Each filesystem entry can also be acted on individually, so long as they are properly ordered.

If you use this model, there is no install/upgrade/uninstall distinction in the core code.  Those APIs can still be presented, but they are used to create the desired state, which is then passed to the resolution engine.  That code can be clear and consistent because it needs only handle one filesystem entry at a time.  Each entry can be properly recorded before it committed to disk, or properly removed before removing the recorded entry.

Additionally, creating an abstraction over the single/multiple file state storage seems like a better way to keep the core code clean and in one place.  The desired state can easily be inspected to determine if it meets the criteria for storing in the registry (contains exactly one directory, file, and link) or not.  The correct interface implementation can be created and not need to have multiple code paths.